### ISSUE TITLE
ComparisonError rpc error: code = Unknown desc = unknown error: max connections reached

### ISSUE THREAD

**Issue by y509319** [argoproj/argo-cd#10263]
**Checklist**:

 - [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.

**Describe the bug:**

I'm having multiple K8S clusters deployed.
Many of them are running with over 50 CRDs each. Syncing with a private BitBucket repository.
On a lot of them, I see errors in the ArgoCD UI apps that state:

`rpc error: code = Unknown desc = unknown error: max connections reached`

Which on its own fails applications sync and flaps them between the Synced and Unknown statuses:

`Updated sync status: Synced -> Unknown`
`Updated sync status: Unknown -> Synced`
`Updated sync status: Synced -> Unknown`

**Logs:**

The messages related to the issue are only seen in the application-controler and repo-server pods.:

argocd-application-controller:
`time="2022-08-09T13:26:39Z" level=info msg="Normalized app spec: {\"status\":{\"conditions\":[{\"lastTransitionTime\":\"2022-08-09T13:26:38Z\",\"message\":\"rpc error: code = Unknown desc = unknown error: max connections reached\",\"type\":\"ComparisonError\"}]}}"`

argocd-repo-server:
`time="2022-08-09T13:26:39Z" level=error msg="finished unary call with code Unknown" error="unknown error: max connections reached" grpc.code=Unknown grpc.method=GenerateManifest grpc.request.deadline="2022-08-09T13:27:38Z" grpc.service=repository.RepoServerService grpc.start_time="2022-08-09T13:26:38Z" grpc.time_ms=946.323 span.kind=server system=grpc`

The only discussion that I've found that could do something with my issue is #3278 from 24th of March 2020, but I'm already using
version 2.0.3 of ArgoCD which was released 27th of May 2021.

How could I see if this has something to do with the MaxIdleConns parameter?
If yes, how could I increase the parameter to a greater value (other than branching)?
If no, what else should I check so that I have some clues what causes this?

**Comment by DesislavaNPetrova** [comment]
Any updates on this? I'm also facing this issue.

**Comment by crenshaw-dev** [comment]
Unfortunately that error message has very little context, so it's difficult to tell what connections are maxing out. 

More recent versions of Argo CD wrap these error messages more effectively to give more context. I'd highly recommend upgrading, both because of that and because there are a large number of security vulnerabilities applicable to 2.0.3.

@DesislavaNPetrova what version are you running? Can you check logs in the application controller and repo-server to see if there are related error or warning messages?

**Comment by DesislavaNPetrova** [comment]
@crenshaw-dev Thanks! We gonna to upgrade and check it again.

**Comment by y509319** [comment]
An update for everyone who might encounter this or similar issue.

Had a discussion with Atlassian support as well.
Turned out we were hitting BitBucket API limits as described in their page here:
[https://support.atlassian.com/bitbucket-cloud/docs/api-request-limits](https://support.atlassian.com/bitbucket-cloud/docs/api-request-limits)

We had to distribute the requests with more authentication keys than what we have ATM.

10x for the assistance. Closing this.
